packages/langium/src/syntax-tree.ts

```diff
@@ -35,6 +35,11 @@ export interface GenericAstNode extends AstNode {
     [key: string]: unknown
 }
 
+export type OptionalAstNode<T extends AstNode> = {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    [P in keyof T]: P extends keyof AstNode ? T[P] : T[P] extends any[] | boolean ? T[P] : T[P] | undefined;
```

Comment:
Thanks!

Would it make sense to make this recursive too? I believe a child node that we want to recur download is either `AstNode[]` or `AstNode | undefined` or `AstNode`. Do I miss anything else?

Here's my attempt to take a stab at it. (This is the first time in my life that I do type-level programming actually! Apologies if anything is off)

```
export type OptionalAstNode<T extends AstNode> = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [P in keyof T]: 
    P extends keyof AstNode ? 
      T[P] : 
      T[P] extends AstNode[] ? 
        OptionalAstNode<T[P][number]>[] : 
        T[P] extends any[] | boolean ? 
          T[P] : 
          T[P] extends AstNode | undefined ? 
            OptionalAstNode<Extract<T[P], AstNode>> | undefined :  
            (T[P] extends AstNode ? OptionalAstNode<T[P]> : T[P]) | undefined;
};
```

Reply:
> Thanks!
> 
> Would it make sense to make this recursive too?

I think it does! 😀